### PR TITLE
Pure Julia WKT2 to PROJJSON conversion 

### PR DESCRIPTION
**Pull request by Omar-Elrefaei**
Working towards #150
/claim #150

Todo:
- [x] Better and consistent error msgs
- [x] Figure out the proper workaround of the `semi_minor_axis` `inverse_flattening` discrepancy (4275, 4267)
  - [x] Elaborate on the `find_diff_paths` workaround.
- [x]  Adjust `test/jsonutils.jl` to project style
- [x] ~Decide whether tests can actually depend on DeepDiffs~ Found better alternative solution.
- [x] Comply with JuliaFormatter 

<details>
<summary>Tools for live development and debugging
</summary>

```julia

# Use DeepDiffs package to view any differences between GDAL's projjson (j1) and our's (j2).
# Differences are shown red/green color. Few insignificant keys are filtered from (j1),
# though there are still more false-negatives than deltapaths filtering
function deepdiffprojjson(j1::J, j2::J) where {J<:Union{Dict}}
  j1 = deepcopy(j1)
  keystodelete = ["bbox", "area", "scope", "usages", "\$schema"]
  for k in keystodelete
    delete!(j1, k)
  end
  try
    delete!(j1["base_crs"], "coordinate_system")
  catch
    nothing
  end
  diff = deepdiff(j1, j2)
  return diff
end

# For live development or debugging.
# run checkprojjson(4275) to see differences between current and expected json output
# The presence of red/green colored output does not neccesarly mean that there is a bug to be fixed.
# If the last printed line is an empty vector, it means the colored diff is likely a superfluous difference.
function checkprojjson(crs::Int; verbose=true)
  gdaljson = gdalprojjsondict(EPSG{crs})
  wktdict = GeoIO.epsg2wktdict(crs)
  jsondict = GeoIO.wkt2json(wktdict)

  # Show pretty-printed WKT if possible
  if verbose && isdefined(Main, :PrettyPrinting)
    @info "Parsed WKT"
    pprintln(wktdict)
  elseif verbose
    @warn "Formatted printing of WKT or JSON is unavailable because PrettyPrinting is not loaded"
  end

  # Show deep diff if possible
  if verbose && isdefined(Main, :DeepDiffs)
    @info "DeepDiff"
    display(deepdiffprojjson(gdaljson, jsondict))
  elseif verbose
    @warn "Detailed colored output is unavailable because DeepDiffs is not loaded."
  end

  diffkeys = deltaprojjson(gdaljson, jsondict)
  @info "JSON keys with a potentially significant difference from expected output:"
  display(diffkeys)
end
```

</details>

### CODE REVIEW

**Comment by juliohm** [comment]
@Omar-Elrefaei this looks promising. Please let me know when it is ready for review.

**Comment by nsajko** [comment]
Some deep issues with the design here:
* Calling `Meta.parse` on the WKT string can't be correct and is obviously a huge hack, no offence. It's parsing foreign code as Julia code!
    * In particular WKT strings have completely different escaping than Julia strings, the former escape by repeating the double quotes, while the latter escape with backslashes.
    * It could also perhaps be considered a security vulnerability, unless the WKT strings are completely vetted upfront on each update of the EPSG data base.
* WKT keywords are case-insensitive, furthermore in some cases several alternative names are possible for the same keyword, so hardcoding the names like here can't be correct.

My design is more complex in part because it tackles these issues and more.

**Comment by juliohm** [comment]
Thank you for sharing @nsajko. Appreciate your inputs.

If you can evolve your PR to a final version to show the technical advantages in practice, we will happily consider it.

Both PRs are still work in progress, but it is really nice to see the high quality of the attempts already.

**Comment by Omar-Elrefaei** [comment]
I think this is functionally mostly there, save for few rare edge cases.
The code itself would take another round of cleanup tho, especially the comments.

few of the concerns brought up by nsajko are justifies, some are design tradeoffs, and some are a non-issue, I think. I'll take time to elaborate soon.

**Comment by juliohm** [comment]
Looking forward to evaluate the pros and cons of each approach. Thank you all for the amazing work and considerations shared so far. It really helps!

**Comment by Omar-Elrefaei** [comment]
@juliohm I think this is ready for review

**Comment by juliohm** [comment]
@Omar-Elrefaei could you please rebase/merge the PR onto the latest `master` branch? I merged another contribution and took the opportunity to better organize some functions in separate files.

File: src/GeoIO.jl [hunk 1]
```diff
@@ -131,6 +131,7 @@ include("extra/off.jl")
 include("extra/ply.jl")
 include("extra/stl.jl")
 include("extra/vtk.jl")
+include("extra/projjson.jl")
```
Comment (juliohm) [inline src/GeoIO.jl:134]: Let's move this file to `src/projjson.jl`

File: test/projjson.schema.json [hunk 2]
```diff
@@ -0,0 +1,1174 @@
+{
```
Comment (juliohm) [inline test/projjson.schema.json:1]: Let's move this file to `test/artifacts/projjson.schema.json`

File: src/extra/projjson.jl [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+  else
+    error("Unimplemented CRS type: $type")
+  end
+end
+
+# process GEOGCRS or GEODCRS nodes, either if they're at top-level or under PROJCRS
+function wkt2json_geog(wkt::Dict)
+  @assert wkt |> keys |> collect |> first in [:GEOGCRS, :GEODCRS]
+  jsondict = Dict{String, Any}()
+  geo_subtype = get_main_key(wkt)
```
Comment (juliohm) [inline src/extra/projjson.jl:57]: Please avoid "_" in variable and function names.

**Review by juliohm — CHANGES_REQUESTED** [review]
Thank you @Omar-Elrefaei for the amazing work here. The code is very readable and that is super appreciated. I noticed that tests are failing. Left a few quick review comments in the meantime.

File: src/GeoIO.jl (hunk 1 above)
Comment (Omar-Elrefaei) [reply src/GeoIO.jl:134]: no problem, but just fyi this is now only called from
```
function projjsonstring(code)
  wktdict = epsg2wktdict(code)
  jsondict = wkt2json(wktdict)
  JSON3.write(jsondict)
end
```
at `src/extras/gis.jl`. I'm not sure if that is the right location for `projjsonstring`.

File: src/extra/projjson.jl (hunk 3 above)
Comment (Omar-Elrefaei) [reply src/extra/projjson.jl]: 👍🏼 many of them is easy.

but what do you recommendnd for the `wkt2json_*` functions?

Comment (juliohm) [reply src/extra/projjson.jl]: I will read the code carefully and get back with possible names 👍🏽

**Comment by juliohm** [comment]
@Omar-Elrefaei could you please eliminate any code that is only used for debugging purposes? I am seeing functions like `debug_json` and variables like `GDALcompat` that are not used in the final tests. Appreciate it if you can cleanup any temporary code and commented lines for final review.

Also, could you please confirm you finished the check boxes you created in the beginning of the PR? Are they still planned?

**Comment by Omar-Elrefaei** [comment]
Sorry Júlio, I think I'll turn this back to a draft for a few days.
I was aware that that code was still not fully ready to be merged, but I thought it might be ready to review the functionality.
I did not communicate that clearly at all. Plus some of my thoughts needed to be put somewhere here (or in comments) to explain a few of the workarounds I took.

**Comment by juliohm** [comment]
No worries, thank you for the through work and attention to detail. Let me know when the PR is ready for review.

**Comment by Omar-Elrefaei** [comment]
Ok I think this is for review.

I've considered adding support for rarely used CRS types to be outside the scope of this PR. To be specific, in the EPSG WKT dataset folder, the following outlines the number of CRSs of each type:
``` bash
$ cd CoordRefSystems/artifacts/epsg-wkt2
$ rg --no-filename --only-matching '^(\w*)' | sort | uniq -c
      4 DERIVEDPROJECTED
     22 ENGCRS
    289 VERTCRS
    453 COMPOUNDCRS
    216 GEODCRS
    825 GEOGCRS
   5289 PROJCRS
```
I've implemented support for the last three `GEODCRS, GEOGCRS, and PROJCRS` EPSGs, with a clear error msg if requested otherwise. All the EPSGs in CoordRefSystems's `get.jl` fall under these three types. Alongside EPSGs in `get.jl`, the testsuite also includes a few more that cover edge cases found during testing with other portions of the dataset.

----------------------------------------------

Since I took a more test-driven approach, comparison testing with GDAL's output was essential for development. Although helpful, there have been annoying cases of "false" negatives. Ie cases where`our_projjson != gdal_projjson` but for all intents and purposes they should be equal. These differences ended up of three categories

1. Optional projjson objects that represent WKT fields that are present in GDAL's WKT but not our WKT
2. Required projjson objects, where the requirement can be satisfied by one of a variety of representations.
  - Often times I'm not sure when ought we use one representation over the other. 
  - Example: in WKT an ellipsoid geometry is only allowed to be defined using the semi_major_axis and inverse_flattening parameters. Whereas in projjson it can by specified similarly, or using the major and minor axes, or only the radius if [inverse_flattening](  https://en.wikipedia.org/wiki/Flattening
) is zero (indicating the ellipsoid is actually a sphere)
3. Floating point discrepancies
  
A significant amount of development time was dedicated to sifting through these false-negatives and developing more tolerant comparison so that true-negatives are prominent whenever there an implementation for a needed projjson object is missing. 

For 1 and 2 the workaround is same, any inequality in the objects in question is ignored. The list of ignored objects is built manually whenever I came across a case that is verified to be a false-negative. This happens in the `deltaprojjson` function.
For 3, I developed alternative code for projjson equality comparison such that for numbers we use `isapprox`, and `isequal` for everything else. 
Note: this comparison code also output the exact location for the sources of inequality in the form of "paths" in the json object (Ie ".datum.ellipsoid.semi_major_axis") This is how the ignoring works for cases 1 and 2.
This function is now called `diffpaths`.

**Comment by juliohm** [comment]
Thank you @Omar-Elrefaei ! I will take a look early morning tomorrow (BRT) 👍🏽

File: test/crsstrings.jl [hunk 4]
```diff
@@ -0,0 +1,35 @@
+@testset "PROJJSON" begin
+  jsonstr = GeoIO.projjsonstring(EPSG{4326})
+  @test isvalidprojjson(JSON3.read(jsonstr))
```
Comment (juliohm) [inline test/crsstrings.jl:3]: Isn't this test already covered in the list below?

File: test/crsstrings.jl [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  
+  epsgcodes = [
+    # https://github.com/JuliaEarth/CoordRefSystems.jl/blob/10b3f944ece7d5c4669eed6dc163ae8d9761afcd/src/get.jl
+    2157, 2193, 3035, 3310, 3395, 3857, 4171, 4207, 4208, 4230, 4231, 4267, 4269, 4274, 4275, 4277,
+    4314, 4326, 4618, 4659, 4666, 4668, 4674, 4745, 4746, 4988, 4989, 5070, 5324, 5527, 8086, 8232,
+    8237, 8240, 8246, 8249, 8252, 8255, 9777, 9782, 9988, 10176, 10414, 25832, 27700, 28355, 29903,
+    # 32662,  # deprecated in 2008, https://github.com/JuliaEarth/CoordRefSystems.jl/issues/262
+    2180, 32600, 32700,
```
Comment (juliohm) [inline test/crsstrings.jl:11]: What are these extra 3 codes?

File: test/crsstrings.jl [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    
+    # WKT strings with non-standard measurement units
+    3407,
+    # WKT projjson with base_crs.datum.prime_meridian
+    31288,
+    # WKT projjson with coordinate_system.axis[1].meridian
+    2986,
```
Comment (juliohm) [inline test/crsstrings.jl:18]: Could you please add a few more sentences in this comment, explaining what exactly is happening with these codes that makes them special? Also, could you please sort the codes?

File: test/testutils.jl [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+# Note: diffpaths uses isapprox to compare numbers to avoid false negatives
+function deltaprojjson(json1, json2; exact=false)
+  diffpaths = finddiffpaths(json1, json2)
+
+  # return full diff in case of exact comparison
+  exact && return diffpaths
+
+  # paths to ignore in approximate comparison
+  # bbox, area, scope, ... are not required to
+  # fully describe the coordinate reference system
```
Comment (juliohm) [inline test/testutils.jl:44]: I've added this comment. Is my understanding correct about these keys? We are ignoring them because they are not required by the PROJJSON schema, right?

**Review by juliohm — CHANGES_REQUESTED** [review]
This is an amazing work as usual Omar. Thank you for the detailed comments and careful analysis of corner cases. It is been a pleasure to work together. I left a few comments. I think we are super close from merging.

File: src/crsstrings.jl [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+  length(keys(d)) == 1 || throw(ArgumentError("Dictionary must have exactly one key."))
+  return first(keys(d))
+end
+
+# From a vector of WKT nodes, find ones that start with `key`
+function finditems(key::Symbol, list::Vector)::Vector{Dict}
+  filter(x -> rootkey(x) == key, list)
+end
+
+function finditem(key::Symbol, list::Vector)::Union{Dict,Nothing}
```
[the remark this answers is not in the document]
Comment (Omar-Elrefaei) [reply src/crsstrings.jl:13]: ......yeah there is no such feature 👀 these is an extra "s" in `finditems` ... seems that my effort at julian names was not very successful 😆

File: test/testutils.jl [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+  isequalfield(v1, v2) ? nothing : [path]
+end
+
+# helper function to compare fields in JSON objects
+isequalfield(x::Number, y::Number) = isapprox(x, y)
+isequalfield(x, y) = isequal(x, y)
+
+# -----------------------------------------
+# Tools for live development and debugging
+# -----------------------------------------
```
[the remark this answers is not in the document]
Comment (Omar-Elrefaei) [reply test/testutils.jl:122]: sure, yeah these can be in the PR here or CONTRIBUTING.md/README.md

File: test/crsstrings.jl (hunk 6 above)
Comment (juliohm) [reply test/crsstrings.jl:18]: Sort numerically, but aside the main list from CRS.jl. It is nice to know that they are not yet supported by CRS.jl, but the current list of 3 codes is not sorted. I believe the EPSG database adds some structure to these integer codes by grouping them into specific value ranges.

File: test/testutils.jl [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+    push!(paths, "$(path)[$(i)]")
+  end
+  return paths
+end
+
+function finddiffpaths(v1, v2, path)
+  isequalfield(v1, v2) ? nothing : [path]
+end
+
+# helper function to compare fields in JSON objects
```
Comment (juliohm) [inline test/testutils.jl:116]: Could you please share an example where this comparison needs to be approximate? I am worried that we might be hitting a bug because these floating point values should match to the last digits. We copy/paste all digits from the EPSG database, and they usually matter.

File: test/crsstrings.jl (hunk 4 above)
Comment (Omar-Elrefaei) [reply test/crsstrings.jl:3]: Yes I just thought all that testing construction is complicated and it might not be obvious how to simply run one test.

File: src/crsstrings.jl (hunk 8 above)
[the remark this answers is not in the document]
Comment (juliohm) [reply src/crsstrings.jl:13]: Oh I missed the "s" at the end. Thanks for clarifying. We just need to remove the type annotation of the output then.

File: test/crsstrings.jl (hunk 4 above)
Comment (juliohm) [reply test/crsstrings.jl:3]: We can remove this one then. The loop with all codes is clear enough 👍🏽

File: test/testutils.jl (hunk 10 above)
Comment (juliohm) [reply test/testutils.jl:116]: When you think you've seen every possible weird thing about GIS, this shows up 🤦🏽‍♂️ 

We should add a special test just for 3407, with a clear comment pointing to the two WKT sources. I understand GDAL is using some old WKT version.

**Comment by juliohm** [comment]
> My question is: **when should we calculate and output `semi_minor_axis`, and when should we save `inverse_flattening` as is?**

That is a tricky one. From what you described, we have the following constraints:

- WKT requires semi_major + inverse_flattening
- PROJJSON requires semi_major + semi_minor

In that case we have no alternative other than converting the second parameter with the formula.

However, I am still confused with this:

> See [4326](https://spatialreference.org/ref/epsg/4326/) for example of being saved as is.

Do you mean that this specific code is one where PROJJSON stores inverse_flattening? I couldn't find the string in the link.

**Comment by Omar-Elrefaei** [comment]
projjson link from the page (
https://spatialreference.org/ref/epsg/4326/projjson.json)


No, the projjson schema allows semi_major+inv_flattening, or
semi_major+semi_minor, or radius.
Schema-wise the three are equal alternatives. That's why the isvalid test
passes.

Radius is acceptable when inv_flattening=0 and thus major=minor=radius
(spec mentions zero as a special value but it is mathematically infinity)


Blessed regards,
Omar Elrefaei

**Comment by Omar-Elrefaei** [comment]
> That is a tricky one. From what you described, we have the following constraints:
> * PROJJSON requires semi_major + semi_minor

No, the PROJJSON schema allows semi_major+inv_flattening, or semi_major+semi_minor, or radius.
Schema-wise the three are equal alternatives. That's why the `isvalid` test passes.

Radius is acceptable when inv_flattening=0 and thus major=minor=radius
(spec mentions 0 as a special value but it is mathematically infinity)

> Do you mean that this specific code is one where PROJJSON stores inverse_flattening? I couldn't find the string in the link.

No, 95% of PROJJSON store inverse_flattening. Everytime I give a code in the discussion here it is just an example, there is usually others. 
Sorry, I gave the general epsg link, here is 4326 PROJJSON link https://spatialreference.org/ref/epsg/4326/projjson.json

File: test/testutils.jl (hunk 10 above)
Comment (Omar-Elrefaei) [reply test/testutils.jl:116]: This is doesn't have to do with 3407 specifically. if you use `isequal` instead of `isapprox` 5 codes fail out 53 (testset shows 106 tests because each code is tested twice). And it seems that the 10% ratio holds for the whole of the dataset.
The testing with `isapprox` was an important part of development.
Note: the floating point discrepancies are never a result of us calculating something, it is always WKT vs WKT. I've tracked down multitude of them during development.  In fact there is NO floating point calculations done in the wkt2json code at all. I once had a `round` call but it has since been removed.

**Comment by juliohm** [comment]
> No, the PROJJSON schema allows semi_major+inv_flattening, or semi_major+semi_minor, or radius.
> No, 95% of PROJJSON store inverse_flattening.

Based on the information you shared above, I can come back to your original question:

> My question is: **when should we calculate and output `semi_minor_axis`, and when should we save `inverse_flattening` as is?**

I think we should just forward `inverse_flattening`? If PROJJSON supports all those alternatives, and this is the only alternative that doesn't introduce conversion errors, it should be the chosen one. Do you mean that GDAL is sometimes picking a different alternative?

File: test/testutils.jl (hunk 10 above)
Comment (juliohm) [reply test/testutils.jl:116]: Got it. If these are all noise in WKT vs. WKT we should be fine with the use of `isapprox` 👍🏽

File: test/crsstrings.jl (hunk 5 above)
Comment (Omar-Elrefaei) [reply test/crsstrings.jl:11]: they are from the very end of CoordRefSystems.jl/src/get.jl, last two are from the loop at the bottom. Note that get.jl has some ESRIs that are excluded here.

File: test/crsstrings.jl [hunk 11]
```diff
@@ -0,0 +1,35 @@
+@testset "PROJJSON" begin
+  jsonstr = GeoIO.projjsonstring(EPSG{4326})
+  @test isvalidprojjson(JSON3.read(jsonstr))
+  
+  epsgcodes = [
+    # https://github.com/JuliaEarth/CoordRefSystems.jl/blob/10b3f944ece7d5c4669eed6dc163ae8d9761afcd/src/get.jl
+    2157, 2193, 3035, 3310, 3395, 3857, 4171, 4207, 4208, 4230, 4231, 4267, 4269, 4274, 4275, 4277,
+    4314, 4326, 4618, 4659, 4666, 4668, 4674, 4745, 4746, 4988, 4989, 5070, 5324, 5527, 8086, 8232,
+    8237, 8240, 8246, 8249, 8252, 8255, 9777, 9782, 9988, 10176, 10414, 25832, 27700, 28355, 29903,
```
[the remark this answers is not in the document]
Comment (Omar-Elrefaei) [reply test/crsstrings.jl:9]: Yes it is. if you want, I'm able to also auto generate that list using: 

```julia-repl
julia> methods(CoordRefSystems.get)
# 177 methods for generic function "get" from CoordRefSystems:
   [1] get(::Type{EPSG{32760}})
...
...
julia> methods(CoordRefSystems.get)[1].sig.parameters[2].parameters[1]
EPSG{32760}
```

**Comment by Omar-Elrefaei** [comment]
> Do you mean that GDAL is sometimes picking a different alternative?

Yes.

This is totally independent from the floating point discussion.

**Comment by nsajko** [comment]
I wonder if testing against PROJ instead of against GDAL might be simpler.

File: test/testutils.jl [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+  # bbox, area, scope, ... are not required to
+  # fully describe the coordinate reference system
+  pathstoignore = ["bbox", "area", "scope", "usages", "\$schema"]
+
+  # occasionally GDAL's ellipsoid doesn't have inverse_flattening and instead has semi_minor_axis
+  # that is calculated from semi_major_axis * (1 - 1/inverse_flattening)
+  push!(pathstoignore, "datum.ellipsoid.semi_minor_axis")
+  push!(pathstoignore, "datum.ellipsoid.inverse_flattening")
+
+  # we don't have the optional node base_crs.coordinate_system in our WKT, in contrast with GDAL
```
[the remark this answers is not in the document]
Comment (Omar-Elrefaei) [reply test/testutils.jl:52]: I guess the language of my comments is confusing because I was trying to express multiple ideas:
It is _NOT_ the case that: "if it is optional, we do not support it"
but more like: "if it is optional _and_ not available for us, we do not support it"


So yes this is additional data that is not required.

**Comment by juliohm** [comment]
I believe we can always use inverse_flattening as you suggested. I'm just
wondering how tests will pass in this case if the GDAL output has something
else. My suggestion was to make the test comparison more evolved for the
ellipsoid parameters but I believe that is not trivial to do given the way
diffpaths simply scans the tree without special cases.

What about we delay this decision to after the other remaining fixes?

Em sex., 23 de mai. de 2025, 16:48, Omar Elrefaei ***@***.***>
escreveu:

**Comment by juliohm** [comment]
Yes, something along your last sentence. If we are producing the wrong
ellipsoid parameters and the tests don't catch that, we might take too long
to uncover the bug.

But if it is something that requires too much more coding, please ignore
this idea for now. We can come back to it in a separate issue.

Em sex., 23 de mai. de 2025, 17:02, Omar Elrefaei ***@***.***>
escreveu:

File: test/crsstrings.jl (hunk 11 above)
[the remark this answers is not in the document]
Comment (juliohm) [reply test/crsstrings.jl:9]: That could be an idea. Let's delay this to the end after we fix the other minor adjustments.

[the remark this answers is not in the document]
Comment (juliohm) [reply test/crsstrings.jl:6]: Let's leave it with an explicit list. We can come back in the future with a more automated method if necessary.

File: test/testutils.jl [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+gdalprojjsondict(code) = JSON3.read(gdalprojjsonstring(code), Dict)
+
+# validate generated json against PROJJSON schema
+function isvalidprojjson(json)
+  path = joinpath(@__DIR__, "artifacts", "projjson.schema.json")
+  schema = Schema(JSON3.parsefile(path))
+  return isvalid(schema, json)
+end
+
+# I can not figure out why the validation test fails without this!
```
[the remark this answers is not in the document]
Comment (juliohm) [reply test/testutils.jl:29]: @Omar-Elrefaei can you give an example where the tests fail without this roundtrip? I think this is the last issue that we need to discuss before the merge.

[the remark this answers is not in the document]
Comment (Omar-Elrefaei) [reply test/testutils.jl:29]: I've investigated this quite a bit when developing and I couldn't get it to make sense. I gave up on it to be honest and just conceded that it is probably some low level anomaly somewhere
the json-dict we produce does not pass the schema validation right away, but it does pass after converting it to json-string and then back to json-dict.  The preplexing thing is that they are both equal
```julia-repl
julia> ourjson = GeoIO.wkt2json(GeoIO.epsg2wktdict(2157))
Dict{String, Any} with 6 entries:
  "name"              => "IRENET95 / Irish Transverse Mercator"
  "base_crs"          => Dict{String, Any}("datum"=>Dict{String, Any}("name"=>"IRENET95", "e…
  "id"                => Dict{String, Any}("authority"=>"EPSG", "code"=>2157)
  "coordinate_system" => Dict{String, Any}("axis"=>Any[Dict{String, Any}("name"=>"Easting", …
  "conversion"        => Dict{String, Any}("name"=>"Irish Transverse Mercator", "parameters"…
  "type"              => "ProjectedCRS"

julia> ourjson |> isvalidprojjson
false

julia> ourjson |> jsonroundtrip |> isvalidprojjson
true

julia> ourjson == ourjson |> jsonroundtrip
true
```

```julia-repl
# It is not some pointer and memory allocation thing
julia> ourjson |> deepcopy |> isvalidprojjson
false

# More confusing. Dict values
julia> ourjson |> values  .== ourjson |> jsonroundtrip |> values
6-element BitVector:
[ 1; 1; 1; 1; 1 ;1]

julia> ourjson |> values  == ourjson |> jsonroundtrip |> values
false

# Dict keys
julia> ourjson |> keys  .== ourjson |> jsonroundtrip |> keys
6-element BitVector:
[ 1; 1; 1; 1; 1 ;1]

julia> ourjson |> keys  == ourjson |> jsonroundtrip |> keys
true

[the remark this answers is not in the document]
Comment (juliohm) [reply test/testutils.jl:29]: Should we open an issue somewhere, or start a thread on Discourse to ask about this MWE? It is really strange indeed.

File: src/crsstrings.jl [hunk 14]
```diff
@@ -0,0 +1,392 @@
...
+
+# ------------------------------------
+# Convert from WKT string to WKT dict
+# ------------------------------------
+
+epsg2wktdict(::Type{EPSG{Code}}) where {Code} = epsg2wktdict(Code)
+
+function epsg2wktdict(code::Int)
+  str = CoordRefSystems.wkt2(EPSG{code})
+  # TODO upstream to CoordRefSystems
```
Comment (juliohm) [inline src/crsstrings.jl:327]: This has been handled in https://github.com/JuliaRegistries/General/pull/131731

File: src/crsstrings.jl [hunk 15]
```diff
@@ -0,0 +1,382 @@
...
+
+function projjson(CRS)
+  try
+    code = CoordRefSystems.code(CRS)
+    jsonstr = projjsonstring(code)
+    json = JSON3.read(jsonstr, Dict)
+    GFT.ProjJSON(json)
+  catch
+    nothing
+  end
```
Comment (juliohm) [inline src/crsstrings.jl:335]: Can we remove this try-catch?

**Comment by juliohm** [comment]
I remember now that this try-catch block is handling the fallback `Cartesian` CRS. We can sort this out in a separate PR later. Let's just focus on the `jsonroundtrip` issue.

File: src/crsstrings.jl (hunk 15 above)
Comment (juliohm) [reply src/crsstrings.jl:335]: I am recalling now that this fallback is handling the `Cartesian` CRS.

**Comment by juliohm** [comment]
Wow! That is annoying! I don't know why the strings are considered
different just because of their types. Maybe a bug in Julia? Maybe expected
behavior that is counter-intuitive?

Thanks for diving into it!

Em seg., 26 de mai. de 2025, 13:35, Omar Elrefaei ***@***.***>
escreveu:

**Comment by juliohm** [comment]
> Peaking at their code, seems that they should use `AbstractString` in one specific place. I'll open an issue.

Thank you! Please link the issue here for future reference.
